core/src/test/java/io/grpc/util/RoundRobinLoadBalancer2Test.java

```diff
@@ -0,0 +1,351 @@
+    verify(oldSubchannel, times(1)).requestConnection();
+
+    assertThat(loadBalancer.getSubchannels()).containsExactly(removedSubchannel,
+        oldSubchannel);
+
+    subchannels2.clear();
+    subchannels2.put(new EquivalentAddressGroup(oldAddr), oldSubchannel);
+    subchannels2.put(new EquivalentAddressGroup(newAddr), newSubchannel);
+
+    List<ResolvedServerInfoGroup> latestServers = Lists.newArrayList(
+        ResolvedServerInfoGroup.builder()
+            .add(new ResolvedServerInfo(oldAddr))
+            .add(new ResolvedServerInfo(newAddr))
+            .build());
+
+    loadBalancer.handleResolvedAddresses(latestServers, affinity);
+
+    verify(newSubchannel, times(1)).requestConnection();
+    verify(removedSubchannel, times(1)).shutdown();
+
+    assertThat(loadBalancer.getSubchannels()).containsExactly(oldSubchannel,
+        newSubchannel);
+
+    verify(mockHelper, times(3)).createSubchannel(any(EquivalentAddressGroup.class),
+        any(Attributes.class));
+    verify(mockHelper, times(2)).updatePicker(pickerCaptor.capture());
+
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterStateChangeBeforeResolution() throws Exception {
+    loadBalancer.handleSubchannelState(mockSubchannel,
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+    verifyNoMoreInteractions(mockSubchannel);
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterStateChangeAndResolutionError() throws Exception {
+    loadBalancer.handleSubchannelState(mockSubchannel,
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+    verifyNoMoreInteractions(mockSubchannel);
+    verifyNoMoreInteractions(mockHelper);
+  }
+
+  @Test
+  public void pickAfterStateChange() throws Exception {
+    when(mockHelper.createSubchannel(any(EquivalentAddressGroup.class), any(Attributes.class)))
+        .then(new Answer<Subchannel>() {
+          @Override
+          public Subchannel answer(InvocationOnMock invocation) throws Throwable {
+            return createMockSubchannel();
+          }
+        });
+    loadBalancer.handleResolvedAddresses(Lists.newArrayList(servers.keySet()), Attributes.EMPTY);
+    verify(mockHelper, times(1)).updatePicker(isA(Picker.class));
+
+    Subchannel subchannel = loadBalancer.getSubchannels().iterator().next();
+    AtomicReference<ConnectivityStateInfo> subchannelStateInfo = subchannel.getAttributes().get(
+        STATE_INFO);
+
+    assertThat(subchannelStateInfo.get()).isEqualTo(ConnectivityStateInfo.forNonError(IDLE));
+
+    loadBalancer.handleSubchannelState(subchannel,
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+
+    assertThat(subchannelStateInfo.get()).isEqualTo(
+        ConnectivityStateInfo.forNonError(ConnectivityState.READY));
+
+    loadBalancer.handleSubchannelState(subchannel,
+        ConnectivityStateInfo.forTransientFailure(Status.UNKNOWN));
```

Comment:
Use a custom status here and verify instance identity below.

Also, verify that for each state change, a picker with expected content is created.